fix(sharding): Update async migration 0004 rollback logic

macobo opened the pull request:
## Problem

See https://github.com/PostHog/posthog/issues/8920. Async migration failed on benchmarking cluster and rollback logic was not correct, leaving it in an incorrect state.


## Changes

- Refactors async migration operations to be easier to introspect via console (used during development) and mock in tests to check rollback logic.
- Changes to rollback logic to make it robust
- Rollback logic tests

Note this does not tackle the underlying error on ATTACH - planning on gathering extra info on this by running the migration more times :)

## How did you test this code?
See tests. Planning on running this on benchmarking cluster again.

macobo committed:
Make SQL async migration operations properly debuggable

@dataclass makes the object have a proper __repr__ with the SQL, instead
of the "custom" debug_context logic.

macobo committed:
ATTACH to REPLACE

macobo committed:
Rollback renaming logic

macobo committed:
Update definitions

macobo committed:
Refactor 0004 async migration: make rollbacks robust

macobo committed:
Make zookeeper paths unique for sharding async migration

yakkomajuri commented on posthog/async_migrations/definition.py:
```diff
@@ -24,50 +25,43 @@ class AsyncMigrationType:
-    @classmethod
-    def get_db_op(
-        cls,
+class AsyncMigrationOperationSQL(AsyncMigrationOperation):
```
👍 for this refactor

macobo replied:
Tests are failing though: the issue is that the behavior of "what should happen if rollback sql is None" was previously unspecified. I think it was automatically supposed to "pass", but can you confirm this?

yakkomajuri replied:
hmm in the past this was discussed and the idea was that unspecified rollbacks should indeed error so that people catch this issue during development, as we'd like to have rollbacks for everything.

The solution would be to explicitly set `rollback=""`. 

We can certainly discuss changing this behavior though

macobo replied:
The APIs here are kind of badly designed in that we're assuming users _need_ to pass in arguments but don't enforce that language-level. There's easy ways to fix that but this will balloon the scope of this PR - splitting them out into a separate PR, will post soon.

yakkomajuri commented on posthog/async_migrations/migrations/0004_replicated_schema.py:
```diff
@@ -143,26 +158,33 @@ def replicated_table_operations(self, table: TableMigrationData):
+            for table_name, create_table_query in table.extra_tables.items():
+                yield AsyncMigrationOperationSQL(sql=create_table_query, rollback=f"DROP TABLE IF EXISTS {table_name}")
+
+        if isinstance(table, ShardedTableMigrationData) and table.materialized_view_name is not None:
```
I believe this lingered? see line 184 - it's the same condition + I believe it's the correct one

macobo replied:
Nope. This is needed for this whole thing to work.

Problem is that materialized views _can_ exist without kafka tables being there, but _cannot_ be created. All of this is a hack to ensure things are done in the right sequence even in a rollback.

yakkomajuri replied:
Right, supposed this could have been the case 👍

yakkomajuri reviewed (changes requested):
mostly looks good - a few comments, one particularly which I think should be addressed (if I'm correct).

I'd like to once again ask you to next time split this up into smaller PRs though. There's a clear refactor here which I can approve in 5 seconds cause it doesn't change functionality and then there are other changes batched together that I need to carefully review.

If you split them reviewing will be easier and faster

macobo commented on posthog/async_migrations/test/test_0004_replicated_schema.py:
```diff
@@ -72,10 +72,36 @@ def test_migration(self):
+
+        migration_successful = start_async_migration(MIGRATION_NAME)
+        self.assertFalse(migration_successful)
+
```
How? We already assert that the migration failed and the other test asserts that the migration succeeds if our patching above wasn't the case.

yakkomajuri replied:
assert the status is `RolledBack` and not `Errored`

macobo committed:
Rename test module to avoid testing-related issues, refactor sql argument

macobo committed:
Fix rollbacks: don't verify schema during get_new_engine calculations

macobo committed:
Fix table_engines typing

macobo committed:
Add is_active predicate

macobo committed:
Fix async migrations testing issue

macobo committed:
Fix predicate

macobo committed:
Rename a test migration

macobo committed:
Prefix tables with database name on ATTACH

Lacking this is causing issues for person_distinct_id2 in the wild

macobo committed:
Handle insanity around MOVE PARTITION

We seem to have hit an ReplicatedMergeTree edge case with
`person_distinct_id2` which requires retrying the move operations a few
times.

To the best of my understanding:
1. `person_distinct_id2` does not have a partition key
2. An unpartitioned table creates a `tuple()` partition lazily when it's referenced for the first time
3. REPLACE PARTITION operation does a zk GET and SET operation
4. For person_distinct_id2 table, the GET is returning that zk node does
not exist, then clickhouse tries to create it whilst it's being created
in the background

It's safe to retry all of this since we're only dropping parts once the
previous move has completed.

macobo commented:
Closing - created follow-up PRs as more refactoring needed to happen.
